src/InvariantTheory/invariant_rings.jl

```diff
@@ -127,16 +127,81 @@ function Base.show(io::IO, IR::InvRing)
 end
 
 function reynolds_molien_via_singular(IR::InvRing)
+   @assert !ismodular(IR)
    if !isdefined(IR, :reynolds_singular) || !isdefined(IR, :molien_singular)
       R = polynomial_ring(IR)
       singular_matrices = _action_singular(IR)
-      rey, mol = Singular.LibFinvar.reynolds_molien(singular_matrices...)
+
+      if iszero(characteristic(coefficient_ring(IR)))
+         rey, mol = Singular.LibFinvar.reynolds_molien(singular_matrices...)
+      else
+         error("Not implemented yet") # See https://github.com/oscar-system/Singular.jl/issues/486
+      end
       IR.reynolds_singular = rey
       IR.molien_singular = mol
    end
    return IR.reynolds_singular, IR.molien_singular
 end
 
+function reynolds_operator(IR::InvRing{FldT, GrpT, T}, f::T) where {FldT, GrpT, T <: MPolyElem}
+   @assert parent(f) === polynomial_ring(IR)
+
+   rey, _ = reynolds_molien_via_singular(IR)
+   fSing = singular_ring(polynomial_ring(IR))(f)
+   fReySing = Singular.LibFinvar.evaluate_reynolds(rey, fSing)
+   # fReySing is an ideal...
+   @assert length(gens(fReySing)) == 1
+   return polynomial_ring(IR)(gens(fReySing)[1])
+end
+
+function molien_series(IR::InvRing)
+   @assert !ismodular(IR)
+   _, mol = reynolds_molien_via_singular(IR)
+   # Singular does not build a new polynomial ring for the univariate Hilbert series
+   # (how could it after all), but uses the first variable of the given ring.
+
+   R = polynomial_ring(IR)
+   K = coefficient_ring(IR)
+   f1 = R(mol[1, 1])
+   g1 = R(mol[1, 2])
+
+   S, t = PolynomialRing(K, "t", cached = false)
+   cs = zeros(K, degree(f1, 1) + 1)
+   for (c, e) in zip(coefficients(f1), exponent_vectors(f1))
+      cs[e[1] + 1] = K(c)
+   end
+   f2 = S(cs)
+   cs = zeros(K, degree(g1, 1) + 1)
+   for (c, e) in zip(coefficients(g1), exponent_vectors(g1))
+      cs[e[1] + 1] = K(c)
+   end
+   g2 = S(cs)
+   return f2//g2
+end
+
+
+function invariant_basis(IR::InvRing, d::Int)
+   @assert d >= 0 "Dimension must be non-negative"
+   R = polynomial_ring(IR)
+   if d == 0
+      return elem_type(R)[ one(R) ]
+   end
+
+   rey, _ = reynolds_molien_via_singular(IR)
+   basisSing = Singular.LibFinvar.invariant_basis_reynolds(rey, d)
+   res = Vector{elem_type(R)}()
+   # I prefer to return an empty array if the vector space is zero dimensional
+   # (and not [ 0 ]), so that the length of the result is always the dimension
+   # of the vector space.
```

Comment:
I don't think this is a matter of preference, but of correctness: `[ 0 ]` is not a basis in any sense I am aware of...